prune: Add JSON support and separate code

aawsome opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------

Separates the pruning code in three functions:

- Find out what to do
- Print statistcs (+ added JSON output here)
- do the actual pruning work

In a next step, unit tests can be added for all three parts (not included here)

This PR conflicts with #2881 - I'll fix this depending on what PR is tackled first.

Was the change discussed in an issue or in the forum before?
------------------------------------------------------------

Refactoring the code: discussion with @fd0 
Also closes #3129 

Checklist
---------

- [x] I have read the [Contribution Guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches)
- [x] I have enabled [maintainer edits for this PR](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)
- [ ] I have added tests for all changes in this PR
- [x] I have added documentation for the changes (in the manual)
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (template [here](https://github.com/restic/restic/blob/master/changelog/TEMPLATE))
- [x] I have run `gofmt` on the code in all commits
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits)
- [ ] I'm done, this Pull Request is ready for review

aawsome committed:
prune: Unify statistics computation

+ prepare JSON output

aawsome committed:
prune: separate collecting/printing/pruning

aawsome committed:
prune: Add JSON output

aawsome commented:
rebased after #3148 was merged.

MichaelEischer commented on cmd/restic/cmd_prune.go:
```diff
@@ -170,7 +171,18 @@ func runPruneWithRepo(opts PruneOptions, gopts GlobalOptions, repo *repository.R
 	}
 
-	return prune(opts, gopts, repo, usedBlobs)
+	removePacksFirst, repackPacks, removePacks, keepBlobs, stats, err :=
```
I'd prefer to move `removePacksFirst, repackPacks, removePacks, keepBlobs` into a struct. That avoids a lot of code churn when we need a new IDSet or similar.

aawsome replied:
Nice idea! Is changed!

MichaelEischer commented on cmd/restic/cmd_prune.go:
```diff
@@ -187,40 +199,54 @@ type packInfoWithID struct {
-// modified in the process.
-func prune(opts PruneOptions, gopts GlobalOptions, repo restic.Repository, usedBlobs restic.BlobSet) error {
-	ctx := gopts.ctx
+type pruneStats struct {
```
I'd like to have a 
```go
MessageType         string  `json:"message_type"` // "summary"
```
field as with the backup command. That should make it easier to extend the output format in the future to e.g. print progress reports

aawsome replied:
Also a very good idea! Is added.

MichaelEischer commented on cmd/restic/cmd_prune.go:
```diff
@@ -422,44 +448,76 @@ func prune(opts PruneOptions, gopts GlobalOptions, repo restic.Repository, usedB
+	// calculate totals for statistics
+	stats.Blobs.Total = stats.Blobs.Used + stats.Blobs.Unused + stats.Blobs.Duplicate
+	stats.Blobs.RemoveTotal = stats.Blobs.Remove + stats.Blobs.RepackRm
+	stats.Blobs.Remain = stats.Blobs.Total - stats.Blobs.Remove - stats.Blobs.RepackRm
```
`stats.Blobs.Total - stats.Blobs.RemoveTotal` would be simpler

aawsome replied:
changed.

MichaelEischer commented on cmd/restic/cmd_prune.go:
```diff
@@ -187,40 +199,54 @@ type packInfoWithID struct {
+// collectPrune selects which files to rewrite and which to delete and which blobs to keep.
+// Also some statistics are returned.
+// The map usedBlobs is modified in the process.
+func collectPrune(opts PruneOptions, gopts GlobalOptions, repo restic.Repository, usedBlobs restic.BlobSet) (removePacksFirst, repackPacks, removePacks restic.IDSet, keepBlobs restic.BlobSet, stats pruneStats, err error) {
```
What about `planPrune` as method name?

aawsome replied:
I like that name and also named the struct `prunePlan` :wink:

MichaelEischer commented on cmd/restic/cmd_prune.go:
```diff
@@ -187,40 +199,54 @@ type packInfoWithID struct {
+		Unref        uint64 `json:"unreferenced"`
+		Total        uint64 `json:"total"`
+		Repack       uint64 `json:"repack"`
+		Repackrm     uint64 `json:"repack_remove"`
```
s/Repackrm/RepackRm/

aawsome replied:
changed.

MichaelEischer reviewed (changes requested):
I was wondering how the JSON should look like to make it easier to extend it to a streaming output as with the backup command in the future. The most important part is probably to add a message type and a comment that this may change in the future.

`restic prune --json` still prints a lot of non-JSON text.

aawsome commented:
@MichaelEischer  I included your proposals and also rebased as #3139 is merged now. The `prunePlan` struct was quite handy ;-)

> `restic prune --json` still prints a lot of non-JSON text.
Yes, so far `--json` can be only used in combination with `--quiet`. I was hoping for #3009 which IMO include the json checks for the various print functions...

aawsome committed:
Add changelog and docu

MichaelEischer commented:
> Yes, so far `--json` can be only used in combination with `--quiet`. I was hoping for #3009 which IMO include the json checks for the various print functions...

Unfortunately, it is not possible to let `--json` imply `--quiet`. For example the `backup` command allows for different levels of verbosity for its JSON output.

aawsome commented:
rebased.

MichaelEischer commented:
> Unfortunately, it is not possible to let `--json` imply `--quiet`. For example the `backup` command allows for different levels of verbosity for its JSON output.

#3009 won't be able to let `--json` imply `--quiet` as that would break the `backup` command output (the format must not force a certain log level). So this has to be handled locally.

aawsome committed:
prune: Remove standard output when using --json

aawsome commented:
> #3009 won't be able to let `--json` imply `--quiet` as that would break the `backup` command output (the format must not force a certain log level). So this has to be handled locally.

Actually I was thinking of something like the last commit I added. IMO this works and doesn't break anything except removing the output for some command that don't support `--json` if it is given.
Also makes many checks for `gopts.JSON` obsolete, but this would be an independent follow-up.

I also rebased to current master.

aawsome commented:
> IMO this works and doesn't break anything

Ok, it breaks `find` which uses the `Print` and `Printf` functions to manually generate JSON output :weary: 

I'll add some more checks for `gopts.JSON` and leave this cleanup to #3009 or a separate PR.

MichaelEischer commented:
I've rebased to PR to fix the merge conflicts.

darkdragon-001 commented:
Thanks for your great work so far!
Are there any updates? I am specifically interested in the JSON output.

darkdragon-001 commented:
@aawsome has this been implemented somewhere else or why was it closed?

darkdragon-001 commented:
I rebased and improved this in #5239.
